### ISSUE TITLE
 "CUDA error: an illegal memory access" with explicit cast to float16 

### ISSUE THREAD

**Issue by tastyminerals** [NVIDIA/apex#580]
I am training a version of unet with joint classification and semantic segmentation using `O1` level. The training crashes after I explicitly cast `box_coord_tensor` in `roi_pool` function.
```python
        rois = roi_pool(
            input=classification_feature_map_tensor, # FLOAT16 
            boxes=box_coord_tensor.half(), # FLOAT32 IF NOT CASTED EXPLICITLY
            output_size=roi_size,
            spatial_scale=1,
        )
```
`classification_feature_map_tensor` comes as float16 since it is handled by amp while `box_coord_tensor` comes from input batch which is float32. However, `roi_pool` requires tensors to have equal precision and throws
```
RuntimeError: Expected tensor for argument #1 'input' to have the same type as tensor for argument #2 'rois'; but type Variable[CUDAHalfType] does not equal Variable[CUDAFloatType] (while checking arguments for ROIPool_forward_cuda) (checkSameType at /pytorch/aten/src/ATen/TensorUtils.cpp:140)
```
So, when I cast #2 tensor to float16, CUDA throws memory access error below.

```
  File "/usr/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/usr/lib/python3.7/site-packages/apex/amp/handle.py", line 123, in scale_loss
    optimizer._post_amp_backward(loss_scaler)
  File "/usr/lib/python3.7/site-packages/apex/amp/_process_optimizer.py", line 241, in post_backward_no_master_weights
    post_backward_models_are_masters(scaler, params, stashed_grads)
  File "/usr/lib/python3.7/site-packages/apex/amp/_process_optimizer.py", line 120, in post_backward_models_are_masters
    scale_override=grads_have_scale/out_scale)
  File "/usr/lib/python3.7/site-packages/apex/amp/scaler.py", line 119, in unscale
    self.unscale_python(model_grads, master_grads, scale)
  File "/usr/lib/python3.7/site-packages/apex/amp/scaler.py", line 89, in unscale_python
    self.dynamic)
  File "/usr/lib/python3.7/site-packages/apex/amp/scaler.py", line 9, in scale_check_overflow_python
    cpu_sum = float(model_grad.float().sum())
RuntimeError: CUDA error: an illegal memory access was encountered
```

**Comment by mcarilli** [comment]
When in doubt, always prefer casting to FP32.  In this case (I think) you're calling into a custom torchvision op that may not have an FP16 implementation.  Cast both inputs to FP32 instead of FP16 and it should work.

**Comment by tastyminerals** [comment]
I casted everything to float32
```python
rois = roi_pool(
    input=classification_feature_map_tensor.float(), 
    boxes=box_coord_tensor.float(),
    output_size=self.roi_size,
    spatial_scale=1,
)
```
The `roi_pool` works and passes but the exception is thrown in apex here
```python
with amp.scale_loss(loss, self.optimizer) as scaled_loss:
    scaled_loss.backward() # exception is thrown
```

**Comment by tastyminerals** [comment]
```
2019-11-04 10:35:43,186 - INFO - __main__ - starting training
2019-11-04 10:35:43,186 - INFO - net.train.trainer - starting training of model, going to train 100 epochs
2019-11-04 10:35:43,190 - INFO - net.train.trainer - running epoch 1
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32768.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8192.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 2048.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 512.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
2019-11-04 10:35:53,378 - INFO - mumo.train.trainer - epoch 1; average train loss nan; processed 10 batches in 10.19 seconds, 1.02 sec per batch on average
2019-11-04 10:35:53,379 - INFO - mumo.train.trainer - epoch 1; starting validation
2019-11-04 10:35:56,085 - INFO - mumo.train.trainer - epoch 1: validation loss nan
2019-11-04 10:35:56,085 - INFO - mumo.train.trainer - epoch 1: validation loss did not decrease, patience left 9
2019-11-04 10:35:56,085 - INFO - mumo.train.trainer - running epoch 2
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16.0
(...)
  File "/usr/lib/python3.7/site-packages/apex/amp/handle.py", line 123, in scale_loss
    optimizer._post_amp_backward(loss_scaler)
  File "/usr/lib/python3.7/site-packages/apex/amp/_process_optimizer.py", line 182, in post_backward_with_master_weights
    models_are_masters=False)
  File "/usr/lib/python3.7/site-packages/apex/amp/scaler.py", line 119, in unscale
    self.unscale_python(model_grads, master_grads, scale)
  File "/usr/lib/python3.7/site-packages/apex/amp/scaler.py", line 89, in unscale_python
    self.dynamic)
  File "/usr/lib/python3.7/site-packages/apex/amp/scaler.py", line 9, in scale_check_overflow_python
    cpu_sum = float(model_grad.float().sum())
RuntimeError: CUDA error: an illegal memory access was encountered
```

**Comment by tastyminerals** [comment]
Could it be related to this

>2.1.10. GEMM Algorithms Numerical Behavior
Some GEMM algorithms split the computation along the dimension K to increase the GPU occupancy, especially when the dimension K is large compared to dimensions M and N. When this type of algorithm is chosen by the cuBLAS heuristics or explicitly by the user, the results of each split is summed deterministically into the resulting matrix to get the final result.
For the routines cublas<t>gemmEx and cublasGemmEx, when the compute type is greater than the output type, the sum of the split chunks can potentially lead to some intermediate overflows thus producing a final resulting matrix with some overflows. Those overflows might not have occured if all the dot products had been accumulated in the compute type before being converted at the end in the output type.
This computation side-effect can be easily exposed when the computeType is CUDA_R_32F and Atype, Btype and Ctype are in CUDA_R_16F.

**Comment by mcarilli** [comment]
I don't think it's running out of memory.  With O1, for the backward pass (https://github.com/NVIDIA/apex/issues/580#issuecomment-549171867) does it error on the very first backward pass?  And what is the exception trace that is thrown?

**Comment by tastyminerals** [comment]
Correct, with `O1` it fails on first backward pass.
```
CUDA_LAUNCH_BLOCKING=1 python train.py --config-file config/config.gin --log-level INFO                                                                                                 
2019-11-04 19:29:08,258 - INFO - __main__ - setting random seed to 42
2019-11-04 19:29:08,258 - INFO - __main__ - setting up train data
2019-11-04 19:29:08,264 - INFO - __main__ - split data with valid fraction 0.2 --> # train data: 40, # valid data: 10
2019-11-04 19:29:08,268 - INFO - net.utils.class_weights - calculating class weights with c=1.04 for box weights and c=1.04 for segmentation weights
2019-11-04 19:29:16,816 - INFO - net.utils.class_weights - calculated box class weights: tensor([ 1.5608, 21.2831, 22.9914, 16.3494, 23.2191, 21.6754, 25.2760, 25.3858,
        23.1732, 25.0054, 19.9499, 10.7810, 19.6184, 20.9051])
2019-11-04 19:29:16,817 - INFO - net.utils.class_weights - calculated segmentation class weights: tensor([0.0821, 0.1714, 0.1662, 0.1396, 0.1677, 0.1864, 0.1912, 0.2489, 0.1080])
2019-11-04 19:29:16,832 - INFO - __main__ - setting up loss function
2019-11-04 19:29:16,832 - INFO - __main__ - combining loss by sum with box loss weight 1.0 and segmentation loss weight 1.0
2019-11-04 19:29:16,832 - INFO - __main__ - setting up model
2019-11-04 19:29:16,891 - INFO - __main__ - setting up trainer instance
Selected optimization level O1:  Insert automatic casts around Pytorch functions and Tensor methods.

Defaults for this optimization level are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic
Processing user overrides (additional kwargs that are not None)...
After processing overrides, optimization options are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic
Warning:  multi_tensor_applier fused unscale kernel is unavailable, possibly because apex was installed without --cuda_ext --cpp_ext. Using Python fallback.  Original ImportError was: ImportError('/usr/lib/python3.7/site-packages/amp_C.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _ZN3c105ErrorC1ENS_14SourceLocationERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE')
2019-11-04 19:29:22,263 - INFO - __main__ - starting training
2019-11-04 19:29:22,263 - INFO - mumo.train.trainer - starting training of model, going to train 100 epochs
2019-11-04 19:29:22,263 - INFO - net.train.trainer - running epoch 1
...
  File "train.py", line 267, in train
    trained_model_state, optimizer_state, metrics = trainer.train()
  File "/home/user/net/train/trainer.py", line 132, in train
    scaled_loss.backward()
  File "/home/user/.local/lib/python3.7/site-packages/torch/tensor.py", line 150, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/home/user/.local/lib/python3.7/site-packages/torch/autograd/__init__.py", line 99, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasGemmEx( handle, opa, opb, m, n, k, &falpha, a, CUDA_R_16F, lda, b, CUDA_R_16F, ldb, &fbeta, c, CUDA_R_16F, ldc, CUDA_R_32F, CUBLAS_GEMM_DFALT_TENSOR_OP)`
```

**Comment by tastyminerals** [comment]
According to docs `CUBLAS_STATUS_EXECUTION_FAILED` means "the function failed to launch on the GPU". I wonder what could be the possible reasons for that since it was launched on GPU several times before it crashes.

**Comment by ptrblck** [comment]
@tastyminerals Are you using variable input sizes, i.e. are some inputs larger than others?
If so, could it be related to [this issue](https://github.com/pytorch/pytorch/issues/24018)?
If you are using CUDA10.0, could you update to 10.1, please, and check, if it's working?

**Comment by tastyminerals** [comment]
No, the input sizes do not change but we have multi-modal input, that is:
```
chargrid_tensor [2 x 65 x 512 x 368]  # batch_size x char_vocab x h x w
box_coordinates [595 x 4] # word_boxes x coords
```
Cuda and cudnn versions: cuda 10.1.243-1, cudnn 7.6.4.38-1

The model structure is the following:
```
BoxClassificationSemanticSegmentation(
  (backbone): UNetWideOneEncoderOneDecoder(
    (conv_encode_1): Sequential(
      (0): Conv2d(65, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): ReLU()
      (2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (3): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (4): ReLU()
      (5): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
      (7): ReLU()
    )
    (conv_encode_2): Sequential(
      (0): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): ReLU()
      (2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (3): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (4): ReLU()
      (5): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
      (7): ReLU()
    )
    (conv_encode_3): Sequential(
      (0): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): ReLU()
      (2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (3): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (4): ReLU()
      (5): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
      (7): ReLU()
    )
    (bottleneck): Sequential(
      (0): Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): ReLU()
      (2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (3): Conv2d(512, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (4): ReLU()
      (5): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (conv_decode_3): Sequential(
      (0): Conv2d(512, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): ReLU()
      (2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (3): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (4): ReLU()
      (5): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (6): ConvTranspose2d(256, 128, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), output_padding=(1, 1))
      (7): ReLU()
    )
    (conv_decode_2): Sequential(
      (0): Conv2d(256, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): ReLU()
      (2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (3): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (4): ReLU()
      (5): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (6): ConvTranspose2d(128, 128, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), output_padding=(1, 1))
      (7): ReLU()
    )
    (conv_decode_1): Sequential(
      (0): Conv2d(192, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): ReLU()
      (2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (3): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (4): ReLU()
      (5): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (6): ConvTranspose2d(128, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), output_padding=(1, 1))
      (7): ReLU()
    )
  )
  (box_classification_head): BoxClassificationHeadHighway(
    (box_classifier): Sequential(
      (0): Highway(
        (proj): Linear(in_features=1024, out_features=1024, bias=True)
        (transform): Linear(in_features=1024, out_features=1024, bias=True)
      )
      (1): Tanh()
      (2): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (3): Linear(in_features=1024, out_features=14, bias=True)
      (4): BatchNorm1d(14, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
  )
  (semantic_segmentation_head): SemanticSegmentationHead1Conv(
    (conv): Conv2d(64, 9, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  )
)
```

So, I did light model pruning like reducing it to box classification task only leaving single `cross_entropy` loss, that is `u-net -> cross_entropy`. No changes.

**Comment by someAdjectiveNoun** [comment]
I get a similar error with the forward pass. After some epochs, it gives the following error.

`RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)``

**Comment by tastyminerals** [comment]
I managed train the model without crashing (at least reach 5 and 6 epoch) with `batch_size=1` and `O2` opt-level. Anything else leads to exception.
```
batch_size=1, opt-level=O1 --> crashes after couple of epochs
batch_size=1, opt-level=O2 --> works fine
batch_size=1, opt-level=O3 --> crashes after couple of epochs

batch_size=2, opt-level=O1 --> crashes after couple of epochs
batch_size=2, opt-level=O2 --> crashes after couple of epochs
batch_size=2, opt-level=O3 --> crashes after couple of epochs
```
Unfortunately, even though with `O2` I am able to train the loss is still `nan` right after the first epoch :( I have `cuda 10.1.243-2` installed.

**Comment by tastyminerals** [comment]
> @tastyminerals Are you using variable input sizes, i.e. are some inputs larger than others?
> If so, could it be related to [this issue](https://github.com/pytorch/pytorch/issues/24018)?
> If you are using CUDA10.0, could you update to 10.1, please, and check, if it's working?

I cannot reproduce the bug, the code below works fine on my machine.
```
torch.zeros((16*2**20 - 512)//2 + 1, 1, dtype=torch.float16, device='cuda:0') @ torch.zeros(1, 2, dtype=torch.float16, device='cuda:0')
```

**Comment by ptrblck** [comment]
@tastyminerals @someAdjectiveNoun 
Could you try to post a (small) code snippet to reproduce this issue?

**Comment by someAdjectiveNoun** [comment]
> @tastyminerals @someAdjectiveNoun
> Could you try to post a (small) code snippet to reproduce this issue?

The problem is solved now. How? The problem was actually caused by using [BioBERT](https://github.com/dmis-lab/biobert) model that I used. Using the BERT in Pytorch works smoothly. The problem seems to be coming from BioBERT.

**Comment by tastyminerals** [comment]
> @tastyminerals @someAdjectiveNoun
> Could you try to post a (small) code snippet to reproduce this issue?

Unfortunately, you'd require custom dataset which we cannot share.

**Comment by tastyminerals** [comment]
I updated APEX and reran the experiments. Now, previously working `batch_size=1, opt-level=O2` stopped working and crashes right after the first epoch. However, now there are some useful debug messages.

```
Traceback (most recent call last):

  File "train.py", line 339, in main
    train()
  File "/home/pavel/.local/lib/python3.7/site-packages/gin/config.py", line 1073, in gin_wrapper
    utils.augment_exception_message_and_reraise(e, err_str)
  File "/home/pavel/.local/lib/python3.7/site-packages/gin/utils.py", line 49, in augment_exception_message_and_reraise
    six.raise_from(proxy.with_traceback(exception.__traceback__), None)
  File "<string>", line 3, in raise_from
  File "/home/pavel/.local/lib/python3.7/site-packages/gin/config.py", line 1050, in gin_wrapper
    return fn(*new_args, **new_kwargs)
  File "train.py", line 269, in train
    trained_model_state, optimizer_state, metrics = trainer.train()
  File "/home/pavel/dev/gini/multi-modal-ner/mumo/train/trainer.py", line 127, in train
    scaled_loss.backward()
  File "/home/pavel/miniconda3/envs/gini_torch/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/home/pavel/.local/lib/python3.7/site-packages/apex/amp/handle.py", line 127, in scale_loss
    should_skip = False if delay_overflow_check else loss_scaler.update_scale()
  File "/home/pavel/.local/lib/python3.7/site-packages/apex/amp/scaler.py", line 200, in update_scale
    self._has_overflow = self._overflow_buf.item()
RuntimeError: CUDA error: an illegal memory access was encountered
```

**Comment by tastyminerals** [comment]
Here is the chunk of training code.
```python
        for epoch in range(1, self.num_epochs + 1):

            logger.info(f"running epoch {epoch}")

            avg_train_loss = 0
            epoch_start_time = timeit.default_timer()

            # set model to training mode, validation switches some things like dropout off
            self.model.train()

            for step, sample_batch in enumerate(self.train_data, start=1):
                sample_batch = self._sample_to_device(sample_batch)
                self.optimizer.zero_grad()

                doc_id_batch = sample_batch[DOC_ID]
                logits_dict = self.model(sample_batch)
                loss = self.criterion(logits_dict, sample_batch)

                logger.debug(
                    f"epoch {epoch}: step {step}; loss {loss.item()}; doc ids {doc_id_batch}"
                )

                with amp.scale_loss(loss, self.optimizer) as scaled_loss:
                    scaled_loss.backward()

                self.optimizer.step()

                avg_train_loss += loss.item()

            epoch_end_time = timeit.default_timer()
            epoch_time = epoch_end_time - epoch_start_time

            avg_train_loss /= len(self.train_data)  # <-- crashes here
```

**Comment by jbartolozzi** [comment]
> I pulled recent **apex** master and reran the experiments. Now, previously working `batch_size=1, opt-level=O2` stopped working and crashes right after the first epoch.
> However, now there are some useful debug messages.
> 
> With `O1`:
> 
> ```
> Traceback (most recent call last):
> 
>   File "train.py", line 339, in main
>     train()
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/config.py", line 1073, in gin_wrapper
>     utils.augment_exception_message_and_reraise(e, err_str)
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/utils.py", line 49, in augment_exception_message_and_reraise
>     six.raise_from(proxy.with_traceback(exception.__traceback__), None)
>   File "<string>", line 3, in raise_from
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/config.py", line 1050, in gin_wrapper
>     return fn(*new_args, **new_kwargs)
>   File "train.py", line 269, in train
>     trained_model_state, optimizer_state, metrics = trainer.train()
>   File "/home/pavel/dev/gini/multi-modal-ner/mumo/train/trainer.py", line 127, in train
>     scaled_loss.backward()
>   File "/home/pavel/miniconda3/envs/gini_torch/lib/python3.7/contextlib.py", line 119, in __exit__
>     next(self.gen)
>   File "/home/pavel/.local/lib/python3.7/site-packages/apex/amp/handle.py", line 127, in scale_loss
>     should_skip = False if delay_overflow_check else loss_scaler.update_scale()
>   File "/home/pavel/.local/lib/python3.7/site-packages/apex/amp/scaler.py", line 200, in update_scale
>     self._has_overflow = self._overflow_buf.item()
> RuntimeError: CUDA error: an illegal memory access was encountered
> ```
> 
> With `O3`:
> 
> ```
>   File "train.py", line 343, in <module>
>     main()
>   File "/home/pavel/.local/lib/python3.7/site-packages/click/core.py", line 764, in __call__
>     return self.main(*args, **kwargs)
>   File "/home/pavel/.local/lib/python3.7/site-packages/click/core.py", line 717, in main
>     rv = self.invoke(ctx)
>   File "/home/pavel/.local/lib/python3.7/site-packages/click/core.py", line 956, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/home/pavel/.local/lib/python3.7/site-packages/click/core.py", line 555, in invoke
>     return callback(*args, **kwargs)
>   File "train.py", line 339, in main
>     train()
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/config.py", line 1073, in gin_wrapper
>     utils.augment_exception_message_and_reraise(e, err_str)
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/utils.py", line 49, in augment_exception_message_and_reraise
>     six.raise_from(proxy.with_traceback(exception.__traceback__), None)
>   File "<string>", line 3, in raise_from
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/config.py", line 1050, in gin_wrapper
>     return fn(*new_args, **new_kwargs)
>   File "train.py", line 269, in train
>     trained_model_state, optimizer_state, metrics = trainer.train()
>   File "/home/pavel/dev/gini/multi-modal-ner/mumo/train/trainer.py", line 131, in train
>     avg_train_loss += loss.item()
> RuntimeError: CUDA error: an illegal memory access was encountered
>   In call to configurable 'train' (<function train at 0x7f0a8829b840>)
> ```
> 
> Prepending `CUDA_LAUNCH_BLOCKING=1`
> 
> ```
>   File "train.py", line 339, in main
>     train()
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/config.py", line 1073, in gin_wrapper
>     utils.augment_exception_message_and_reraise(e, err_str)
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/utils.py", line 49, in augment_exception_message_and_reraise
>     six.raise_from(proxy.with_traceback(exception.__traceback__), None)
>   File "<string>", line 3, in raise_from
>   File "/home/pavel/.local/lib/python3.7/site-packages/gin/config.py", line 1050, in gin_wrapper
>     return fn(*new_args, **new_kwargs)
>   File "train.py", line 269, in train
>     trained_model_state, optimizer_state, metrics = trainer.train()
>   File "/home/pavel/dev/gini/multi-modal-ner/mumo/train/trainer.py", line 127, in train
>     scaled_loss.backward()
>   File "/home/pavel/.local/lib/python3.7/site-packages/torch/tensor.py", line 166, in backward
>     torch.autograd.backward(self, gradient, retain_graph, create_graph)
>   File "/home/pavel/.local/lib/python3.7/site-packages/torch/autograd/__init__.py", line 99, in backward
>     allow_unreachable=True)  # allow_unreachable flag
> RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasGemmEx( handle, opa, opb, m, n, k, &falpha, a, CUDA_R_16F, lda, b, CUDA_R_16F, ldb, &fbeta, c, CUDA_R_16F, ldc, CUDA_R_32F, CUBLAS_GEMM_DFALT_TENSOR_OP)`
>   In call to configurable 'train' (<function train at 0x7fd08eb2e840>)
> ```

I'm getting the same results trying to run pix2pixHD training on Quadro RTX 6000

**Comment by tastyminerals** [comment]
@jbartolozzi Quadro RTX 6000 has like 24GB of GPU memory? ... good lord. Did you try to use different batch sizes? Does it crash with `batch_size = 1`? I had a feeling it has something to do with memory capacity but if your model crashes, I don't know what could it be.

**Comment by jbartolozzi** [comment]
With opt-level=00 there's no crashing.
These results are with a batch size of 1.

**Comment by tastyminerals** [comment]
Yeah, the `opt-level=O0` doesn't crash because it does not modify the model in any way. It's is a dry run. But looks like this ticket won't be solved in near future.

**Comment by tripzero** [comment]
as with @jbartolozzi I have tried pix2pixHD with cuda 10.2 and getting the same results.

**Comment by mcarilli** [comment]
https://github.com/NVIDIA/apex/issues/580#issuecomment-549299087 may be fixed by https://github.com/pytorch/pytorch/pull/37569.  The fix has been in master for a while, but did not make 1.5.1.

I still recommend moving to [torch.cuda.amp](https://github.com/NVIDIA/apex/issues/818).  However, if the above PR is the right diagnosis, the problem is not in apex, but in Pytorch's FP16 gemv implementation, so you'll have to update Pytorch whether you choose apex or `torch.cuda.amp`.

**Comment by tripzero** [comment]
> [#580 (comment)](https://github.com/NVIDIA/apex/issues/580#issuecomment-549299087) may be fixed by [pytorch/pytorch#37569](https://github.com/pytorch/pytorch/pull/37569). The fix has been in master for a while, but did not make 1.5.1.
> 
> I still recommend moving to [torch.cuda.amp](https://github.com/NVIDIA/apex/issues/818). However, if the above PR is the right diagnosis, the problem is not in apex, but in Pytorch's FP16 gemv implementation, so you'll have to update Pytorch whether you choose apex or `torch.cuda.amp`.

May try pytorch nightly builds as it look like 1.6 is just around the corner...

**Comment by tripzero** [comment]
Just tried pytorch 1.7.0 nightly.  While I didn't get a CUBLAS_STATUS_EXECUTION_FAILED error, I did get the "Gradient overflow" and my GAN started producing black images :|.

**Comment by mcarilli** [comment]
Make sure you're following the guidance for [multiple models/losses/optimizers](https://pytorch.org/docs/master/notes/amp_examples.html#working-with-multiple-models-losses-and-optimizers).  (`retain_graph` in that snippet is present because the two backward passes share some graph sections, it has nothing to do with amp.  You may not need `retain_graph` for your own multi-model network.)

An example GAN training-loop step with proper `torch.cuda.amp` control flow can be found [here](https://colab.research.google.com/github/pytorch/ignite/blob/master/examples/notebooks/CycleGAN_with_torch_cuda_amp.ipynb#scrollTo=vrJls4p-FRcA&line=93&uniqifier=1), courtesy of @vfdev-5 (https://twitter.com/pytorch_ignite/status/1262721636844920832).

If that doesn't work, file an issue with a minimal repro on Pytorch github and tag me.

**Comment by seovchinnikov** [comment]
Got the same tracktrace as above (RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasGemmEx( handle, opa, opb, m, n, k, &falpha, a, CUDA_R_16F, lda, b, CUDA_R_16F, ldb, &fbeta, c, CUDA_R_16F, ldc, CUDA_R_32F, CUBLAS_GEMM_DFALT_TENSOR_OP)`)

With last nightly build (1.7.0.dev20200709), Cuda V10.1.243, apex master (https://github.com/NVIDIA/apex/commit/1ff54b8fed441c39dac181091b44fecdca31a403) and cudnn 7.6.3_0 seems to be working fine (using apex API)

**Comment by tripzero** [comment]
@mcarilli converted my training loop to use torch.cuda.amp instead of apex.  It runs... but it doesn't seem like there's any indication that it's actually using 16 bit floats.  Memory usage is identical as non-fp16 as is the speed.  Do you know if there a way to verify amp is working with fp16 correctly?

Here's my modified code from pix2pixHD:

```python
    amp_scaler = GradScaler(enabled=opt.fp16)
    with autocast(enabled=opt.fp16):

            ############## Forward Pass ######################
            losses, generated = model(Variable(data['label']), inst_map,
                                      Variable(data['image']), Variable(data['feat']), infer=save_fake)

            # sum per device losses
            losses = [torch.mean(x) if not isinstance(x, int)
                      else x for x in losses]
            loss_dict = dict(zip(model.module.loss_names, losses))

            # calculate final loss scalar
            loss_D = (loss_dict['D_fake'] + loss_dict['D_real']) * 0.5
            loss_G = loss_dict['G_GAN'] + \
                loss_dict.get('G_GAN_Feat', 0) + loss_dict.get('G_VGG', 0)

       ############### Backward Pass ####################
       # update generator weights
       optimizer_G.zero_grad()
       amp_scaler.scale(loss_G).backward()
        amp_scaler.step(optimizer_G)
        # if opt.fp16:
        #    with amp.scale_loss(loss_G, optimizer_G) as scaled_loss:
        #        scaled_loss.backward()
        # else:
        #    loss_G.backward()
        # optimizer_G.step()

        # update discriminator weights
        optimizer_D.zero_grad()
        amp_scaler.scale(loss_D).backward()
        amp_scaler.step(optimizer_D)

        amp_scaler.update()
```

**Comment by Jmennius** [comment]
We had a similar issue with another cuBLAS API (cublasSgemm()), although @anjani-dhrangadhariya also [eperienced this](https://github.com/NVIDIA/apex/issues/580#issuecomment-551195778).

CUDA Toolkit 11.1 release notes mention an issue fixed in cuBLAS:
> Fixed an issue that caused an Address out of bounds error when calling `cublasSgemm()`.

We had `cublasSgemm()` failing with `CUBLAS_STATUS_EXECUTION_FAILED` for us when built with 10.0 and running on Ampere GPU (3060 Ti). It ran fine on older GPUs (Pascal, Turing).
We had it run successfully on Ampere when we build it with CUDA 11.2.

Basically - try building against the newest CUDA Toolkit available and see if it helps.

P.S. this was with another framework/project, but should still be relevant.
P.P.S. related issue in PyTorch pytorch/pytorch#29795.

**Comment by ShoufaChen** [comment]
Great thanks to this [answer](https://discuss.pytorch.org/t/runtimeerror-cuda-error-cublas-status-invalid-value-when-calling-cublassgemm-handle-opa-opb-m-n-k-alpha-a-lda-b-ldb-beta-c-ldc/124544/6?u=alpha).

There may be a mismatch between the dimension of your input tensor and the dimensions of your `nn.Linear`.
